fml/gpu_thread_merger.h

```diff
@@ -0,0 +1,59 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FML_SHELL_COMMON_TASK_RUNNER_MERGER_H_
+#define FML_SHELL_COMMON_TASK_RUNNER_MERGER_H_
+
+#include "flutter/fml/macros.h"
+#include "flutter/fml/memory/ref_counted.h"
+#include "flutter/fml/message_loop_task_queues.h"
+
+namespace fml {
+
+class MessageLoopImpl;
+
+enum class GpuThreadStatus { kRemainsMerged, kRemainsUnmerged, kUnmergedNow };
+
+class GpuThreadMerger : public fml::RefCountedThreadSafe<GpuThreadMerger> {
+ public:
+  // Merges the GPU thread into platform thread for the duration of
+  // the lease term. Lease is managed by the caller by either calling
+  // |ExtendLease| or |DecrementLease|.
+  // When the caller merges with a lease term of say 2. The threads
+  // are going to remain merged until 2 invocations of |DecreaseLease|,
+  // unless an |ExtendLease| gets called.
+  void MergeWithLease(size_t lease_term);
+
+  void ExtendLease(size_t lease_term);
```

Comment:
> The guarantee of this method to the caller is that the threads will stay merged for at least the lease_term they call with. If another caller tries to reduce the current lease term, we should still have the bigger lease term. In the existing implementation we do not ever call this with a smaller lease term than before, but I think of it as extending as we would never want to lower the lease term.

I can imagine a world where we want to reduce the lease after disposing all platform views.
Anyway not a big deal, I just thought it will be a little easier to follow as `SetLease` but I'm ok with keeping it this way if that is your preference, I still find it a little confusing e.g `ExtendLease(3)` sounds to me like we're extending the current lease by 3, if you want to keep current behavior maybe make it `ExtendLeaseTo(3)`

Reply:
I like `ExtendLeaseTo` I see your point about `SetLease` but i'm worried about setting the lease to a lower time and cause unintended consequences. Changed it.